src/ngx_http_lua_timer.c

```diff
@@ -52,24 +68,154 @@ static u_char *ngx_http_lua_log_timer_error(ngx_log_t *log, u_char *buf,
 static void ngx_http_lua_abort_pending_timers(ngx_event_t *ev);
 
 
+static 
+ngx_inline unsigned long bit_check(const unsigned long pos) {
+    if (pos >= reftable->max_refs) return 1;
+    return reftable->open_refs[pos / BPL] & (1UL << (pos % BPL));
+}
+
+
+static 
+ngx_inline void bit_set(const unsigned long pos)
+{
+    if (pos >= reftable->max_refs) return;
+    unsigned idx = pos / BPL;
+    reftable->open_refs[pos / BPL] |= 1UL << (pos % BPL);
+    if (reftable->open_refs[pos/BPL] == ~0UL) {
+        reftable->full_refs_bits[idx/BPL] |= (1UL << (idx % BPL));
+    }
+}
+
+
+static 
+ngx_inline void bit_clean(const unsigned long pos)
+{
+    if (pos >= reftable->max_refs) return;
+    unsigned idx = pos / BPL;
+    reftable->open_refs[pos / BPL] &= ~(1UL << (pos % BPL));
+    reftable->full_refs_bits[idx / BPL] &= ~(1UL << (idx % BPL));
+}
+
+
+static 
+long find_first_zero_bit(const unsigned long start)
+{
+    if (!bit_check(start)) return start;
+    unsigned _start = start / BPL / BPL;
+    int idx = -1;
+    for (; _start * BPL * BPL < reftable->max_refs;
+          _start++) {
+        if (reftable->full_refs_bits[_start] != ~0UL) {
+            idx = ffz(reftable->full_refs_bits[_start]);
+            break;
+        }
+    }
+    if (idx == -1) return -1;
+    idx = ffz(reftable->open_refs[_start * BPL + idx]) + 
+          (idx + _start * BPL) * BPL;
+
+    if ((unsigned)idx >= reftable->max_refs) return -1;
+    return idx;
+}
+
+
+static
+void expand_reftable(ngx_log_t *log) {
+    if (!reftable) { /* init */
+        reftable = ngx_alloc(sizeof(struct ngx_lua_timer_reftable), log);
```

Comment:
should check NULL

Reply:
I don't get your point, isn't it what `if (!reftable) { ` do to check NULL ?

Reply:
`ngx_alloc` may return `NULL`, which leads to null pointer dereference

Reply:
Yes you right, add `nomem` check.